providers/common/ciphers/aes_gcm.c

```diff
@@ -0,0 +1,565 @@
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        sz = aes_gcm_tls_init(ctx, p->data, p->data_size);
+        if (sz == 0) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_INVALID_AAD);
+            return 0;
+        }
+        ctx->tls_aad_pad_sz = sz;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (aes_gcm_tls_iv_set_fixed(ctx, p->data, p->data_size) == 0) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+static int aes_gcm_stream_update(void *vctx, unsigned char *out, size_t *outl,
+                                 size_t outsize, const unsigned char *in,
+                                 size_t inl)
+{
+    PROV_AES_GCM_CTX *ctx = (PROV_AES_GCM_CTX *)vctx;
+
+    if (outsize < inl) {
+        PROVerr(PROV_F_AES_GCM_STREAM_UPDATE, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return -1;
+    }
+
+    if (aes_gcm_cipher_internal(ctx, out, outl, in, inl) <= 0) {
+        PROVerr(PROV_F_AES_GCM_STREAM_UPDATE, PROV_R_CIPHER_OPERATION_FAILED);
+        return -1;
+    }
+    return 1;
+}
+
+static int aes_gcm_stream_final(void *vctx, unsigned char *out, size_t *outl,
+                                size_t outsize)
+{
+    PROV_AES_GCM_CTX *ctx = (PROV_AES_GCM_CTX *)vctx;
+    int i;
+
+    i = aes_gcm_cipher_internal(ctx, out, outl, NULL, 0);
+    if (i <= 0)
+        return 0;
+
+    *outl = 0;
+    return 1;
+}
+
+static int aes_gcm_cipher(void *vctx,
+                          unsigned char *out, size_t *outl, size_t outsize,
+                          const unsigned char *in, size_t inl)
+{
+    PROV_AES_GCM_CTX *ctx = (PROV_AES_GCM_CTX *)vctx;
+
+    if (outsize < inl) {
+        PROVerr(PROV_F_AES_GCM_CIPHER, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return -1;
+    }
+
+    if (aes_gcm_cipher_internal(ctx, out, outl, in, inl) <= 0)
+        return -1;
+
+    *outl = inl;
+    return 1;
+}
+
+#ifdef FIPS_MODE
+/*
+ * See SP800-38D (GCM) Section 8 "Uniqueness requirement on IVS and keys"
+ *
+ * See also 8.2.2 RBG-based construction.
+ * Random construction consists of a free field (which can be NULL) and a
+ * random field which will use a DRBG that can return at least 96 bits of
+ * entropy strength. (The DRBG must be seeded by the FIPS module).
+ */
+static int aes_gcm_iv_generate(PROV_AES_GCM_CTX *ctx, int offset)
+{
+    int sz = ctx->ivlen - offset;
+
+    /* Must be at least 96 bits */
+    if (sz <= 0 || ctx->ivlen < AES_GCM_IV_DEFAULT_SIZE)
+        return 0;
+
+    /* Use DRBG to generate random iv */
+    if (RAND_bytes(ctx->iv + offset, sz) <= 0)
```

Comment:
Use `rand_byes_ex` in FIPS_MODE (we should disable RAND_bytes in FIPS mode)

Reply:
so I have to store the ossl_ctx when the newctx() happens right?